custom_components/tesla_custom/sensor.py

```diff
@@ -1,112 +1,455 @@
+
+        # If we don't have anything, just return None.
+        if charge_rate is None:
+            return charge_rate
+
+        if self.unit_of_measurement == "km/hr":
+            charge_rate = round(
+                convert(charge_rate, LENGTH_MILES, LENGTH_KILOMETERS), 2
             )
-        except AttributeError:
-            return None
+
+        return charge_rate
 
     @property
     def extra_state_attributes(self):
-        """Return the state attributes of the device."""
-        attr = self._attributes.copy()
-        if self.tesla_device.type == "charging rate sensor":
-            attr.update(
-                {
-                    "time_left": self.tesla_device.time_left,
-                    "added_range": self.tesla_device.added_range,
-                    "charge_energy_added": self.tesla_device.charge_energy_added,
-                    "charge_current_request": self.tesla_device.charge_current_request,
-                    "charge_current_request_max": self.tesla_device.charge_current_request_max,
-                    "charger_actual_current": self.tesla_device.charger_actual_current,
-                    "charger_voltage": self.tesla_device.charger_voltage,
-                    "charger_power": self.tesla_device.charger_power,
-                }
+        """Return device state attributes."""
+        added_range = self._car.charge_miles_added_ideal
+
+        if self._car.gui_range_display == "Rated":
+            added_range = self._car.charge_miles_added_rated
+
+        if self.unit_of_measurement == "km/hr":
+            added_range = round(
+                convert(added_range, LENGTH_MILES, LENGTH_KILOMETERS), 2
             )
-        return attr
+
+        data = {
+            "time_left": self._car.time_to_full_charge,
+            "added_range": added_range,
+            "charge_energy_added": self._car.charge_energy_added,
+            "charge_current_request": self._car.charge_current_request,
+            "charge_current_request_max": self._car.charge_current_request_max,
+            "charger_actual_current": self._car.charger_actual_current,
+            "charger_voltage": self._car.charger_voltage,
+            "charger_power": self._car.charger_power,
+            "charger_phases": self._car.charger_phases,
+            "charge_limit_soc": self._car.charge_limit_soc,
+        }
+        self.attrs.update(data)
+        return self.attrs
+
+
+class TeslaCarChargerEnergy(TeslaCarEntity, SensorEntity):
+    """Representation of a Tesla car energy added sensor."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        car: TeslaCar,
+        coordinator: TeslaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize energy added entity."""
+        super().__init__(hass, car, coordinator)
+        self.type = "energy added"
+        self._attr_device_class = SensorDeviceClass.ENERGY
+        self._attr_state_class = SensorStateClass.TOTAL
```

Comment:
I changed this to `TOTAL` to fix #222. However, it may be worth trying to find a creative solution to create an actual `TOTAL_INCREASING` state class sensor based on some use cases. Basically we would have to reset the value to 0 when a decreasing value is detected...

Reply:
It is odd, it seems to reset to zero at strange times. Logically you would think it would reset to zero after each charge is disconnected, but no.

The teslafi approach to classify charging sessions makes more intuitive sense. Sometimes a longer charging session can be broken into smaller sessions.

The main usecase for this is to keep a running total of energy consumption to populate energy dashboard so you can know how much consumption for a day/ week/ month/ year.

So a pseudo TOTAL_INCREASING sensor would be welcome.

Reply:
For sure. I’m thinking we will just essentially ignore what Tesla is reporting and set 0 when “charging_state” is not “Charging”.

Reply:
Just pushed an update that does this. I'll be keeping an eye on it to see if it's behaving as expected, which should be when `charging_state` is not `Charging`, it returns a value of 0 for `energy_added`.

Reply:
It looks like `charge_energy_added` is still being returned even though `charging_state` is not `Charging`. 

You can see from the screenshot below, it should have only been reporting the `charge_energy_added` when the charging sensor was "On".

<img width="963" alt="Screen Shot 2022-09-10 at 11 31 32 AM" src="https://user-images.githubusercontent.com/32663154/189497069-1c441cf1-2e8a-4fb3-bbb6-b46bae0e60be.png">